easybuild/tools/extract.py

```diff
@@ -0,0 +1,248 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+This module provides the extract_archive command
+and it's helper functions.
+
+This will help you to extract archives in pure python, but
+
+@author: Jens Timmerman (Ghent University)
+"""
+import bz2
+import os
+import copy
+import gzip
+import operator
+import tarfile
+import zipfile
+
+from easybuild.tools import run
+
+from vsc.utils.missing import get_subclasses
+from vsc.utils import fancylogger
+
+_log = fancylogger.getLogger('filetools', fname=False)
+
+BUFFERSIZE = 10240
+
+
+def get_extractor(filename):
+    """Returns the extractor if the file is an archive, or None otherwise
+    """
+    if not filename:
+        return None
+    for cls in get_subclasses(Extractor):
+        if cls.can_handle(filename):
+            return cls
+    return None
+
+
+def extract_archive(filename, destination_dir):
+    """Extracts a given archive to the destination directory"""
+    if not os.path.isfile(filename):
+        _log.debug("Can't extract file %s: no such file" % filename)
+        return None
+
+    if not os.path.isdir(destination_dir):
+        try:
+            os.makedirs(destination_dir)
+        except OSError, err:
+            _log.exception("Can't extract file %s: directory %s can't be created: %err ", filename, destination_dir, err)
+
+    destination_dir = os.path.abspath(destination_dir)
+
+    # recursive extracting please
+    # if we get a .tar.gz we should return the output dir
+    cls = get_extractor(filename)
+    if cls:
+        out = cls.extract(filename, destination_dir)
+        out2 = extract_archive(out, destination_dir)
+        return out2 or out
+    # not an archive
+    return None
+
+
+class Extractor(object):
+    """"This is an abstract implementation for an extractor class
+    it's purpose is to extract compressed archives
+    """
+    # magic number identifying files this class can handle
+    # you can find magic numbers in from http://www.garykessler.net/library/file_sigs.html
+    # e.g.,
+    # LZW: "\x1F\x9D",
+    # LZH: "\x1F\xA0",
+    MAGIC = "This is not a magic numbers string"
+    # MAGIC starts at offset
+    OFFSET = 0
+
+    @classmethod
+    def can_handle(cls, filename):
+        """Returns True if this class can handle the given file
+        This uses magic numbers"""
+        f = open(filename)
+        f.seek(cls.OFFSET)
+        can_handle = f.read(len(cls.MAGIC)) == cls.MAGIC
+        f.close()
+        return can_handle
+
+    @classmethod
+    def extract(cls, filename, destination):
+        """
+        Do the actual extraction
+        This is an abstract method, it will return the result of _extract
+        which has to be implemented in a class extending Extractor
+        """
+        outfile = os.path.join(destination, filename)
+        if filename.endswith(cls.EXTENTION):
```

Comment:
yes, what else should it be?
I've seen tarred files, that do not end in  .tar, 
extentions are just a convention, not everybody follows them, that's the whole point of doing the magic numbers thing, check what it is instead of relying on the convention.